File: src/main/java/com/aoranzhang/ezrentback/service/CityService.java
package com.aoranzhang.ezrentback.service;

import com.aoranzhang.ezrentback.data.entity.City;
import com.aoranzhang.ezrentback.data.repository.CityRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.Set;

@Service
public class CityService {

    @Autowired
    private CityRepository cityRepository;

    @Transactional
    public void createNewCity(City city) {
        if(cityRepository.findByName(city.getName()) != null) {
            cityRepository.findByName(city.getName());
        }else {
            cityRepository.save(city);
        }
    }

    public Set<City> getAllCities() {
        Set<City> cities = new HashSet<>();
        cityRepository.findAll().forEach((City c) -> cities.add(c));
        return cities;
    }

    public City findCityByName(String name) {
        return cityRepository.findByName(name);
    }

    public City getCityById(String ID) {
        return cityRepository.findOne(ID);
    }
}

File: src/main/kotlin/com/aoranzhang/ezrentback/service/CityService.kt
package com.aoranzhang.ezrentback.service

import com.aoranzhang.ezrentback.data.entity.City
import com.aoranzhang.ezrentback.data.repository.CityRepository
import com.aoranzhang.ezrentback.util.JavaToKotlin
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class CityService {

    @Autowired
    private val cityRepository: CityRepository? = null

    val allCities: Set<City>
        get() {
            val cities = HashSet<City>()
            cityRepository!!.findAll().forEach { c: City -> cities.add(c) }
            return cities
        }

    @Transactional
    fun createNewCity(city: City) {
        if (cityRepository!!.findByName(city.name) != null) {
            cityRepository.findByName(city.name)
        } else {
            cityRepository.save(city)
        }
    }

    fun findCityByName(name: String): City? {
        return cityRepository!!.findByName(name)
    }

    fun getCityById(ID: String): City? {
        return JavaToKotlin.OptionalToNullable(cityRepository!!.findById(ID))
    }
}
